Update SegFormer model card (#40417)
huggingface/transformers: docs/source/en/model_doc/segformer.md

hierarchically structured Transformer encoder which outputs multiscale features. It does not need positional encoding,
thereby avoiding the interpolation of positional codes which leads to decreased performance when the testing resolution
differs from training. 2) SegFormer avoids complex decoders. The proposed MLP decoder aggregates information from
different layers, and thus combining both local attention and global attention to render powerful representations. We
show that this simple and lightweight design is the key to efficient segmentation on Transformers. We scale our
approach up to obtain a series of models from SegFormer-B0 to SegFormer-B5, reaching significantly better performance
and efficiency than previous counterparts. For example, SegFormer-B4 achieves 50.3% mIoU on ADE20K with 64M parameters,
being 5x smaller and 2.2% better than the previous best method. Our best model, SegFormer-B5, achieves 84.0% mIoU on
Cityscapes validation set and shows excellent zero-shot robustness on Cityscapes-C.*

The figure below illustrates the architecture of SegFormer. Taken from the [original paper](https://huggingface.co/papers/2105.15203).

> <img width="600" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/segformer_architecture.png"/>

Author: Added the image

> ```python

Reviewer: ```py
import torch
from transformers import pipeline

pipeline = pipeline(task="image-segmentation", model="nvidia/segformer-b0-finetuned-ade-512-512", torch_dtype=torch.float16)
pipeline("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/pipeline-cat-chonk.jpeg")
```

> from transformers import SegformerImageProcessor, SegformerForSemanticSegmentation

Reviewer: ```py
import requests
from PIL import Image
from transformers import AutoProcessor, AutoModelForSemanticSegmentation

url = "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/pipeline-cat-chonk.jpeg"
image = Image.open(requests.get(url, stream=True).raw)

processor = AutoProcessor.from_pretrained("nvidia/segformer-b0-finetuned-ade-512-512")
model = AutoModelForSemanticSegmentation.from_pretrained("nvidia/segformer-b0-finetuned-ade-512-512")

inputs = processor(images=image, return_tensors="pt")
outputs = model(**inputs)
logits = outputs.logits # shape [batch, num_labels, height, width]
```

outputs = model(**inputs)
logits = outputs.logits  # shape [batch, num_labels, height, width]
```

</hfoption>

</hfoptions>

> Quantization reduces the memory burden of large models by representing the weights in a lower precision. Refer to the [Quantization](../quantization/overview) overview for more available quantization backends.

Reviewer: The model isn't very big so no need to have a quantization example

Author: Gotcha, removed it

The example below uses [bitsandbytes](https://huggingface.co/docs/transformers/main/en/quantization/bitsandbytes) to quantize SegFormer weights to 8-bit.

```python
from transformers import BitsAndBytesConfig, AutoModelForSemanticSegmentation

quantization_config = BitsAndBytesConfig(load_in_8bit=True)
model = AutoModelForSemanticSegmentation.from_pretrained(
    "nvidia/segformer-b0-finetuned-ade-512-512",
    quantization_config=quantization_config
)
```

## Notes

> - SegFormer works with **any input size**, padding inputs to be divisible by `config.patch_sizes`.

Reviewer: ```suggestion
- SegFormer works with **any input size**, padding inputs to be divisible by `config.patch_sizes`.
- The most important preprocessing step is to randomly crop and pad all images to the same size (such as 512x512 or 640x640) and normalize afterwards.
```

> - When preprocessing, be mindful of `do_reduce_labels`:
>   - Some datasets (like ADE20K) don’t include background in the labels → set `do_reduce_labels=True`.
>   - Other datasets do include background → set `do_reduce_labels=False`.

Reviewer: ```suggestion
- Some datasets (ADE20k) uses the `0` index in the annotated segmentation as the background, but doesn't include the "background" class in its labels. The `do_reduce_labels` argument in [`SegformerForImageProcessor`] is used to reduce all labels by `1`. To make sure no loss is computed for the background class, it replaces `0` in the annotated maps by `255`, which is the `ignore_index` of the loss function.

   Other datasets may include a background class and label though, in which case, `do_reduce_labels` should be `False`.
```

> - Model variants differ in size and accuracy (MiT-B0 to MiT-B5). Example:
>   - **SegFormer-B4**: 50.3% mIoU on ADE20K with 64M parameters, 5x smaller and better than prior methods.
>   - **SegFormer-B5**: 84.0% mIoU on Cityscapes validation set, strong zero-shot robustness.

Reviewer: ```suggestion
```
